Azure/sonic-buildimage
# [arista] Add DPB for Arista 7050 QX32

This change introduces dynamic port breakout (DPB) for Arista 7050 QX32
model by adding a new SKU suffixed with `-Flex`.

The breakout configuration allowed is the same as in mainline Arista
EOS, i.e. 24 first ports are allowed to be used in 4x10G in addition to
the default 40G mode. The last 8 ports are fixed to 40G. This is due to
ASIC limitations of a total of 104 max ports.

**NOTE**: As described in https://github.com/aristanetworks/sonic/issues/30#issuecomment-820584113 front panel LEDs are likely not working when operating in breakout mode. It is not clear if the LEDs work correctly in 40G mode as I have not had a chance to physically inspect the switch with this patch.

#### Why I did it

Configuring breakout for switches is a really fundamental feature on this 40G switch.

#### How I did it

I reformatted the BCM configuration to be easier to work with and worked out through trial and error with available documentation a configuration that worked for my Arista 7050 QX32.

I also got some help from Arista themselves in https://github.com/aristanetworks/sonic/issues/30.

#### How to verify it

Build this PR with updated YANG models and systemd version revert. See https://github.com/bluecmd/sonic-buildimage/compare/arista-7050-qx32-dpb-pr...bluecmd:arista-7050-qx32-dpb-pr-test.

Enable the new SKU:
```
sudo -i
sonic-cfggen -H -k Arista-7050-QX32-Flex > /etc/sonic/config_db.json
reboot
```

Enable port breakout:
```
sudo config int breakout Ethernet64 4x10G
```

Verify the status:
```
admin@localhost:~$ show int status
  Interface            Lanes    Speed    MTU    FEC         Alias    Vlan    Oper    Admin            Type    Asym PFC
-----------  ---------------  -------  -----  -----  ------------  ------  ------  -------  --------------  ----------
  Ethernet0  125,126,127,128      40G   9100    N/A   Ethernet1/1  routed    down       up             N/A         N/A
  Ethernet4  121,122,123,124      40G   9100    N/A   Ethernet2/1  routed    down       up             N/A         N/A
  Ethernet8      13,14,15,16      40G   9100    N/A   Ethernet3/1  routed    down       up             N/A         N/A
 Ethernet12       9,10,11,12      40G   9100    N/A   Ethernet4/1  routed    down       up             N/A         N/A
 Ethernet16      17,18,19,20      40G   9100    N/A   Ethernet5/1  routed    down       up             N/A         N/A
 Ethernet20      21,22,23,24      40G   9100    N/A   Ethernet6/1  routed    down       up             N/A         N/A
 Ethernet24      25,26,27,28      40G   9100    N/A   Ethernet7/1  routed    down       up             N/A         N/A
 Ethernet28      29,30,31,32      40G   9100    N/A   Ethernet8/1  routed    down       up             N/A         N/A
 Ethernet32      37,38,39,40      40G   9100    N/A   Ethernet9/1  routed    down       up             N/A         N/A
 Ethernet36      33,34,35,36      40G   9100    N/A  Ethernet10/1  routed    down       up             N/A         N/A
 Ethernet40      45,46,47,48      40G   9100    N/A  Ethernet11/1  routed    down       up             N/A         N/A
 Ethernet44      41,42,43,44      40G   9100    N/A  Ethernet12/1  routed    down       up             N/A         N/A
 Ethernet48      53,54,55,56      40G   9100    N/A  Ethernet13/1  routed    down       up             N/A         N/A
 Ethernet52      49,50,51,52      40G   9100    N/A  Ethernet14/1  routed    down       up             N/A         N/A
 Ethernet56      69,70,71,72      40G   9100    N/A  Ethernet15/1  routed    down       up             N/A         N/A
 Ethernet60      65,66,67,68      40G   9100    N/A  Ethernet16/1  routed    down       up             N/A         N/A
 Ethernet64               77      10G   9100    N/A  Ethernet17/1  routed      up       up  QSFP+ or later         N/A
 Ethernet65               78      10G   9100    N/A  Ethernet17/2  routed      up       up             N/A         N/A
 Ethernet66               79      10G   9100    N/A  Ethernet17/3  routed    down       up             N/A         N/A
 Ethernet67               80      10G   9100    N/A  Ethernet17/4  routed    down       up             N/A         N/A
 Ethernet68      73,74,75,76      40G   9100    N/A  Ethernet18/1  routed    down       up  QSFP+ or later         N/A
 Ethernet72      93,94,95,96      40G   9100    N/A  Ethernet19/1  routed    down       up             N/A         N/A
 Ethernet76      89,90,91,92      40G   9100    N/A  Ethernet20/1  routed    down       up             N/A         N/A
 Ethernet80  101,102,103,104      40G   9100    N/A  Ethernet21/1  routed    down       up             N/A         N/A
 Ethernet84     97,98,99,100      40G   9100    N/A  Ethernet22/1  routed    down       up             N/A         N/A
 Ethernet88  109,110,111,112      40G   9100    N/A  Ethernet23/1  routed    down       up             N/A         N/A
 Ethernet92  105,106,107,108      40G   9100    N/A  Ethernet24/1  routed    down       up             N/A         N/A
 Ethernet96      61,62,63,64      40G   9100    N/A    Ethernet25  routed    down       up             N/A         N/A
Ethernet100      57,58,59,60      40G   9100    N/A    Ethernet26  routed    down       up             N/A         N/A
Ethernet104      81,82,83,84      40G   9100    N/A    Ethernet27  routed    down       up             N/A         N/A
Ethernet108      85,86,87,88      40G   9100    N/A    Ethernet28  routed    down       up             N/A         N/A
Ethernet112  117,118,119,120      40G   9100    N/A    Ethernet29  routed    down       up             N/A         N/A
Ethernet116  113,114,115,116      40G   9100    N/A    Ethernet30  routed    down       up             N/A         N/A
Ethernet120          5,6,7,8      40G   9100    N/A    Ethernet31  routed    down       up             N/A         N/A
```

You can also verify using `bcmsh` with `ps` and `phy info`.

#### Which release branch to backport (provide reason below if selected)


#### Description for the changelog

Suggestion: `Add Dynamic Port Breakout (DPB) for Arista 7050 QX32`

#### A picture of a cute animal (not mandatory but encouraged)

![image](https://user-images.githubusercontent.com/149442/114929411-0b4cc280-9e34-11eb-8b91-2dea31ab5146.png)
